### ISSUE TITLE
OnChangeCell, OnChangeCells Callbacks Not occurring in some cases

### ISSUE THREAD

**Issue by alaarab** [rowsncolumns/grid#96]
I am experimenting with callbacks, and these are some problematic behaviors I am seeing that prevent being able to persist in data model. 

**OnChangeCell**
- not triggering on undo/redo
- not triggering on copy/paste
- not triggering when dragging (copy) a cell value down
- many console logs: 'No dependencies for  G6'

**OnChangeCells**
- not triggering on undo/redo
- not triggering on copy/paste
- many console logs: 'No dependencies for  G6'

**Comment by rmdort** [comment]
Yep. I will add it to my todo list.

**Comment by rmdort** [comment]
To continue the discussion, here is my plan for these 2 events

`onChangeCell`

Will be called
* When user manually changes a cell value, from the editor

> Will not be called for undo/redo or delete

`onChangeCells`

Only called during these events
* undo/redo
* copy/paste/cut
* cell filling (dragging cells)
* delete

Let me know if this works

**Comment by alaarab** [comment]
That is an interesting plan - in that case it sounds like both callbacks would need to be used together to handle all cases for reliable persisting model data, and that there is no case where both would be fired at the same time.

Originally it made sense to me that both should handle all cases each, the main difference being that multiple callbacks would be called for onChangeCell. However, if what you propose is the result decision, it seems workable.. as long as it is possible to track all changes with the callbacks to be able to persist model data reliably.

**Comment by rmdort** [comment]
If persisting data model is the concern, you can hook into `onChange` which gets called for all sheet change events, be it copy/paste, undo etc. 

The reason i split cell change events to 2, onChangeCell and onChangeCells, is so that users can easily execute side-effects when a single cell change, which happens most often.

I will make sure to fire `onChangeCells` for all cell change events, both when single cell or multiple cells change.

**Comment by alaarab** [comment]
OnChange occurs very frequently (simply navigating, etc), making it less desirable to use for this situation. It is more beneficial to make minimal data calls for lowest overhead, in which case onChangeCells should suffice if it properly fires on the above mentioned events.

**Comment by rmdort** [comment]
Yes. because `activeCell` and `selections` change when user navigates the sheet. So onChange is triggered more often.

If you want to persist sheet properties like name, tabColor etc, you still need to hook into onChange though.

**Comment by alaarab** [comment]
I understand, another shortcoming of onChange in my use case is it only provides the entire sheet object in the callback, and not the exact changes, which is another reason onChangeCell(s) is better for me.. When a cell changes, I am only wanting to save that cell by itself in the database. It is working good, but not firing on undo/redo/delete/drag, so those changes aren't picked up in my app. Upgrading onChangeCells to fire on the additional events like your original proposal would resolve my use case of this.

**Comment by rmdort** [comment]
You can try the new release. onCellChanges gets called for all cell updates.

**Comment by alaarab** [comment]
Still seems like it is glitchy in cases where you copy and paste Multiple, or Undo/Redo multiple. Like it only shows changes for one cell instead of all.

**Comment by alaarab** [comment]
I have made an attempt to resolve the Copy/Paste and Cut/Paste multiple issues in this pull request: https://github.com/rowsncolumns/grid/pull/109

However, this PR does not resolve 'Undo' yet.

**Comment by rmdort** [comment]
Can you tell me how to reproduce undo/redo issue?

**Comment by alaarab** [comment]
Absolutely - The 'main' issue I brought up here is the onChangeCells callback (which cells were changed?) 

**to reproduce**
add 5 rows of data in column A, copy and paste Column A to Column B, C, and D. Run Undo (ctrl Z) and redo (ctrl Z) and inspect the onChangeCells callback. You will see only one cell is returned in many cases. In some cases, more than one cell may show up. 

I dug into this to try to understand the issue. 

**my findings**
The selection that is passed into the onChangeCells callback in onUndo and onRedo is incorrect, and maybe it should not use these selections at all unless it were to correctly select the cells that need to be undid or redid (even if these selections are not within one specific bound). I was logging out the 'patches' and also inspected applyPatches to understand what was happening here. I was going to try to build the cell interfaces myself and pass that into the onChangeCells callback, but when doing this, I noticed some odd cases in the outputs of the patches in some very specific cases which led me to not work on this yet until I understood the behavior more fully. If you want to give it a shot, since you have a deeper understanding, feel free, that would be great! I haven't been able to use this package in my app until some of these things are ironed out.
